lib/rubocop/cop/rspec/expect_in_hook.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Do not use `expect` in hooks such as `before`.
+      #
+      # @example
+      #   # bad
+      #   before do
+      #     expect(something).to eq 'foo'
+      #   end
+      #
+      #   # bad
+      #   after do
+      #     expect_any_instance_of(Something).to receive(:foo)
+      #   end
+      #
+      #   # good
+      #   it do
+      #     expect(something).to eq 'foo'
+      #   end
+      #
+      #   # good
+      #   it do
+      #     expect_any_instance_of(Something).to receive(:foo)
+      #   end
+      class ExpectInHook < Cop
+        MSG = 'Do not use `%<expect>s` in `%<hook>s` hook'.freeze
+        HOOKS = Hooks::ALL.node_pattern_union.freeze
+
+        def on_block(node)
+          hook(node) do |hook_name, body|
+            expect(body) do |expect|
+              add_offense(expect, :selector,
+                          message(expect.method_name, hook_name))
+            end
+          end
+        end
+
+        private
+
+        def message(expect, hook)
+          format(MSG, expect: expect, hook: hook)
+        end
+
+        def_node_matcher :hook, <<-PATTERN
+          (block (send _ $#{HOOKS} ...) _ $!nil)
+        PATTERN
+
+        def_node_search :expect, <<-PATTERN
+          (send nil {:expect :expect_any_instance_of} ...)
```

Comment:
Yeah it would be nice to centralize as much RSpec DSL traversal as possible but I would only ask @pocke to do so if it is in this PR if is an easy change.